src/Common/LRUResourceCache.h

```diff
@@ -0,0 +1,303 @@
+#pragma once
+
+#include <atomic>
+#include <chrono>
+#include <list>
+#include <memory>
+#include <mutex>
+#include <unordered_map>
+#include <unordered_set>
+#include <base/logger_useful.h>
+
+namespace DB
+{
+template <typename T>
+struct TrivailLRUResourceCacheWeightFunction
+{
+    size_t operator()(const T &) const { return 1; }
+};
+
+/*
+ * A resource cache with key index. There is only one instance for every key which is not like the normal resource pool.
+ * Resource cache has max weight capacity and keys size limitation. If the limitation is exceeded, keys would be evicted
+ * by LRU policy.
+ *
+ * acquire and release must be used in pair.
+ */
+template <
+    typename TKey,
+    typename TMapped,
+    typename WeightFunction = TrivailLRUResourceCacheWeightFunction<TMapped>,
+    typename HashFunction = std::hash<TKey>>
+class LRUResourceCache
+{
+public:
+    using Key = TKey;
+    using Mapped = TMapped;
+    using MappedPtr = std::shared_ptr<Mapped>;
+    class MappedHolder
+    {
+    public:
+        ~MappedHolder()
+        {
+            cache->release(key);
+                
+        }
+        Mapped & value()
+        {
+            return *(val.get());
+        }
+        static bool tryRemove(std::unique_ptr<MappedHolder> * holder_ptr)
```

Comment:
Why would we need such method? It seems redundant to me.

Reply:
In our case, when a remote file is updated, we need to drop the current cached file

Reply:
Ok, but then may be it makes sense to make it's implementation such that if refcount is 0, it removes the key, if non-zero refcount, then it sets the flag in cell `expired = true`. 
So if `expired == true`, then 
1. `get` for this key will return false
2. the weight of value if this key is still taken into account when we check remaining space
3. when `release` is called for this key and refcount gets to 0, then we remove this key from cache

Also I see no reason why it should accept a holder, and not just key. If we have a holder and then decide that corresponding key needs to be removed, then better to get rid of holder first. I mean no need in such method in holder to me.

Reply:
done